src/codec.rs

```diff
@@ -125,78 +154,126 @@ fn number_of_digits(mut n: usize) -> usize {
+                                    // Otherwise, successfully set content_len from the parsed "Content-Length" value
+                                    Ok(content_len) => {
+                                        self.content_len = Some(content_len);
+                                    }
+                                },
+                            }
+                        }
+                    }
                 }
-
-                match code {
-                    ErrorKind::Digit | ErrorKind::MapRes => return Err(ParseError::InvalidLength),
-                    ErrorKind::Char | ErrorKind::IsNot => return Err(ParseError::InvalidType),
-                    _ => return Err(ParseError::MissingHeader),
-                }
-            }
-        };
-
-        let result = match msg {
-            Err(err) => Err(err.into()),
-            Ok(msg) if msg.is_empty() => Ok(None),
-            Ok(msg) => {
-                trace!("<- {}", msg);
-                match serde_json::from_str(msg) {
-                    Ok(parsed) => Ok(Some(parsed)),
-                    Err(err) => Err(err.into()),
+                // No errors occurred during parsing yet but no complete set of headers were parsed
+                Ok(httparse::Status::Partial) => return Ok(None),
+                // An error occurred during parsing of the headers
+                Err(error) => {
+                    self.http_error = Some(error);
                 }
             }
-        };
+        }
 
-        src.advance(len);
-        self.remaining_msg_bytes = 0;
+        // If headers were found but there was an error parsing the correct values (e.g., "Content-Length")
+        if let (Some(headers_len), Err(err)) = (self.headers_len, parse) {
+            // Advance beyond the found headers
+            src.advance(headers_len);
+            // Reset the codec state
+            self.reset();
+            // Return the error
+            return Err(err);
+        }
 
-        result
-    }
-}
+        // "Content-Length" has been parsed
+        let data =
+            if let (Some(headers_len), Some(content_len)) = (self.headers_len, self.content_len) {
+                let delta = headers_len + content_len;
 
-fn parse_message(input: &[u8]) -> IResult<&[u8], &[u8]> {
-    let content_len = delimited(tag("Content-Length: "), digit1, crlf);
+                // Source doesn't contain the full content yet so return and wait for more input
+                if src.len() < delta {
+                    return Ok(None);
+                }
 
-    let utf8 = alt((tag("utf-8"), tag("utf8")));
-    let charset = tuple((char(';'), space0, tag("charset="), utf8));
-    let content_type = tuple((tag("Content-Type:"), is_not(";\r"), opt(charset), crlf));
+                // Parse the JSON-RPC message bytes as JSON
+                let message = &src[headers_len..delta];
+                let message = std::str::from_utf8(message)?;
+
+                log::trace!("<- {}", message);
```

Comment:
Since `trace!()` is already imported, we might want to remove the `log::` prefix to eliminate warnings.